### PR TITLE
feat(ulid): port /x/ulid module

### PR DESCRIPTION
**Pull request by lino-levan**
Closes https://github.com/denoland/deno_std/issues/3522.

I think this is valuable, particularly for Deno KV.

### CODE REVIEW

**Review by iuioiua — CHANGES_REQUESTED** [review]
We probably only want to export the following from `mod.ts`:
* `decodeTime()`
* `factory()`
* `monotonicFactory()`
* `PRNG`
* `ULID`
I think the others should be moved to a `_util.ts` or similar file.

**Comment by iuioiua** [comment]
We also probably want to include the examples from the README in the JSDocs.

File: ulid/test.ts
Comment (iuioiua) [inline ulid/test.ts:16]: Nit: I don't see the benefit of having this extra nesting level with tests. It could be more flat if done as separate `Deno.test()`s. Not a strong opinion, though.

**Review by iuioiua — APPROVED** [review]
Nice!

File: ulid/mod.ts
Comment (iuioiua) [inline ulid/mod.ts:65]: This line is causing the CI errors. Perhaps, just use `detectPrng()`?

File: ulid/mod.ts
Comment (iuioiua) [inline ulid/mod.ts:99]: Ditto

**Review by iuioiua — APPROVED** [review]
LGTM again

**Comment by kt3k** [comment]
@iuioiua Do you know the usage of `monotonicFactory`? I'm not sure what it's used for.

I'm also unsure we need to export `factory`. I guess nobody would replace pseudo random number generator.

I'd suggest we should only expose `ulid` and `decodeTime` functions (and related types) at least for first pass.

**Comment by iuioiua** [comment]
I think `monotonicFactory()` is worth keeping. It's used for generating monotonically increasing ULIDs, which would be valuable in guaranteeing no conflicts when high numbers of ULIDs are generated within the same millisecond.

I agree with the rest of your points.

**Comment by kt3k** [comment]
Ok. Monotonicity looks useful. ([This doc](https://github.com/ulid/spec#monotonicity) also mentions about guaranteeing of generation order).

While monotonicity is useful, `monotonicFactory` looks still redundant as it is for specifying alternative random number generator. I'd suggest we should replace it with non-factory version (such as `ulidMonotonic` or maybe `generateMonotonic`)

**Comment by iuioiua** [comment]
Yeah, it seems a little funny. It only increments monotonically if the last timestamp is less than or equal to the current timestamp (now). I think it's fine how it is with the name `monotonicFactory`. However, that's not a strong opinion.

Half-related: I think deviating from the ULID's JavaScript implementation is fine. Otherwise, we'd be putting ourselves into a box regarding capability. E.g. I found a possible tweak that could improve the performance of ULID generation by 10x while maintaining cryptographically random strings. Though, I'd have to validate it's correctness first 🤣

**Comment by iuioiua** [comment]
WDYT, @kt3k?

**Comment by lino-levan** [comment]
I would almost flip the naming to have `ulid` be monotonically increasing and some sort of non-monotonic `ulid` exist on the side? It feels like outside of migration, everyone should really use monotonic `ulid`s.

**Comment by kt3k** [comment]
With monotonic version, the random bits are not random anymore, and it seems having far higher possibility of collision in distributed usage. I think monotonic version is not something we should recommend by default.

**Comment by lino-levan** [comment]
> With monotonic version, the random bits are not random anymore

I missed that. This makes sense then. Completely agree that no one will replace the PRNG function. I'll rework the PR tonight if I can.

**Comment by iuioiua** [comment]
Oh, I completely missed that you suggested removing the random number generator parameter. Yes, that parameter looks mostly useless to me.

File: ulid/_util.ts
Comment (iuioiua) [inline ulid/_util.ts:56]:
```suggestion
export function encodeRandom(len: number): string {
  let str = "";
  const randomBytes = crypto.getRandomValues(new Uint8Array(len));
  for (let i = 0; i < len; i++) {
    str += ENCODING[randomBytes[i] % ENCODING_LEN];
  }
  return str;
}
```
This function is 10x faster, retains cryptographic randomness and removes the need for the `prng` parameter. Doing this might need tweaking elsewhere in the implementation.

Comment (lino-levan) [reply ulid/_util.ts:56]: I'll run some statistical tests tomorrow, but this makes sense to me.

Comment (lino-levan) [reply ulid/_util.ts:56]: This sort of breaks a lot of tests unfortunately. I'll try to pick this up on the bus tomorrow.

**Comment by lino-levan** [comment]
Return of the flaky tests I guess. This is ready for review @iuioiua.

**Review by iuioiua — CHANGES_REQUESTED** [review]
Cool to see that change implemented! Few things:
* I don't think `monotonicFactory()` should have an `encodeRand` parameter or any parameter.
* This implementation deviates from the original JS implementation (e.g. faster `encodeRandom()` and no `prng` parameter. I think these differences should be pointed out in the JSDocs somewhere.

File: ulid/_util.ts
Comment (iuioiua) [inline ulid/_util.ts:32]:
```suggestion
  if (!Number.isInteger(now) || now < 0 || now > TIME_MAX) {
    throw new Error("Time must be a positive integer less than " + TIME_MAX);
  }
```
Nit

File: ulid/_util.ts
Comment (iuioiua) [inline ulid/_util.ts:9]:
```suggestion
```
I believe this is no longer needed. Right?

File: ulid/_util.ts
Comment (iuioiua) [inline ulid/_util.ts:78]: How come this is here instead of `mod.ts`?

File: ulid/_util.ts
Comment (iuioiua) [inline ulid/_util.ts:99]: How come this is here instead of `mod.ts`?

File: ulid/_util.ts
Comment (iuioiua) [inline ulid/_util.ts:81]:
```suggestion
 * Generates a monotonically increasing ULID.
```

**Review by iuioiua — CHANGES_REQUESTED** [review]

**Comment by lino-levan** [comment]
> I don't think monotonicFactory() should have an encodeRand parameter or any parameter.

Strongly disagree. It is impossible to test without this parameter. It is not exposed to end users so I don't think this matters.

> This implementation deviates from the original JS implementation (e.g. faster encodeRandom() and no prng parameter. I think these differences should be pointed out in the JSDocs somewhere.

Again, not exposed to the user so I don't know if that's relevant.

File: ulid/_util.ts
Comment (lino-levan) [reply ulid/_util.ts:90]: See above.

File: ulid/_util.ts
Comment (lino-levan) [reply ulid/_util.ts:69]: If I'm reading Yoshiya's comments correctly, he doesn't believe these should be exposed to end users. I have to agree. If they are unable to change the PRNG there is no use for this. If it was up to me I'd remove the factory functions entirely but it makes testing for monotonic harder.

Comment (lino-levan) [reply ulid/_util.ts]: I've gone ahead and removed this factory function entirely. I've kept the monotonic factory for now.

File: ulid/_util.ts
Comment (iuioiua) [inline ulid/_util.ts:58]: Is this line reachable? If not, it should be removed.

File: ulid/_util.ts
Comment (iuioiua) [reply ulid/_util.ts]: Nice

File: ulid/mod.ts
Comment (iuioiua) [inline ulid/mod.ts:66]: I totally misread your initial suggestion, @kt3k. Now, I understand, and I like it 😎

**Review by iuioiua — APPROVED** [review]
Regardless of my previous unreachable line comment, LGTM! This implementation is actually better than the original JavaScript implementation because:
1. Random strings are generated 10x faster.
2. The redundant `prng` parameter is removed. Fewer options are better.
3. Monontic ULIDs are possible without having first to create the factory.

Good stuff, @lino-levan and @kt3k!

File: ulid/_util.ts
Comment (lino-levan) [reply ulid/_util.ts:58]: This was supposed to be reachable, but due to a bug it wasn't. Did not affect behavior in any way, but nice catch!

**Comment by iuioiua** [comment]
Anything else pending for this PR?

**Comment by lino-levan** [comment]
Nothing pending afaik. Waiting for any comments from @kt3k.

**Comment by iuioiua** [comment]
I've noticed a bug with `monotonicUlid()` in https://github.com/denoland/saaskit/pull/514. For some reason, ULIDs aren't respecting their `seedTime` parameters. However, `ulid()` works fine.

**Comment by lino-levan** [comment]
Could you write a reproduction? I suspect it may be a feature but I'm not too sure. @kt3k may know more about it.

**Comment by iuioiua** [comment]
Yeah, I'll get a repro snippet when I can.

**Comment by lino-levan** [comment]
Alright I've been staring at this code for too long. I don't see this bug. I'd appreciate a fresh set of eyes.

**Comment by iuioiua** [comment]
Sorry, I haven't had the chance to look into this further. @kt3k, would `std/ulid` be considered an unstable module once merged? If so, I think it'd be worth merging as-is.

**Review by kt3k — APPROVED** [review]
LGTM. Thank you for your efforts! @lino-levan 

Note: Core team also looked in favor of including this.

**Comment by kt3k** [comment]
> would std/ulid be considered an unstable module once merged?

Yes, we are planning to keep ulid unstable (versioned as 0.x) for a while.

**Pull request merged**
